common/txmgr/inmemory_store.go

```diff
@@ -0,0 +1,1869 @@
+			return false
+		}
+		meta := map[string]interface{}{}
+		if err := json.Unmarshal(json.RawMessage(*tx.Meta), &meta); err != nil {
+			return false
+		}
+		if v, ok := meta[metaField].(string); ok {
+			return v == metaValue
+		}
+
+		return false
+	}
+	txsLock := sync.Mutex{}
+	txs := []*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]{}
+	wg := sync.WaitGroup{}
+	ms.addressStatesLock.RLock()
+	defer ms.addressStatesLock.RUnlock()
+	for _, as := range ms.addressStates {
+		wg.Add(1)
+		go func(as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) {
+			for _, tx := range as.FetchTxs(states, filterFn) {
+				etx := ms.deepCopyTx(tx)
+				txsLock.Lock()
+				txs = append(txs, etx)
+				txsLock.Unlock()
+			}
+			wg.Done()
+		}(as)
+	}
+	wg.Wait()
+
+	return txs, nil
+}
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) FindTxesWithMetaFieldByStates(ctx context.Context, metaField string, states []txmgrtypes.TxState, chainID *big.Int) ([]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], error) {
+	if ms.chainID.String() != chainID.String() {
+		return nil, fmt.Errorf("find_txes_with_meta_field_by_states: %w", ErrInvalidChainID)
+	}
+
+	filterFn := func(tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) bool {
+		if tx.Meta == nil {
+			return false
+		}
+		meta := map[string]interface{}{}
+		if err := json.Unmarshal(json.RawMessage(*tx.Meta), &meta); err != nil {
+			return false
+		}
+		if _, ok := meta[metaField]; ok {
+			return true
+		}
+
+		return false
+	}
+
+	txsLock := sync.Mutex{}
+	txs := []*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]{}
+	wg := sync.WaitGroup{}
+	ms.addressStatesLock.RLock()
+	defer ms.addressStatesLock.RUnlock()
+	for _, as := range ms.addressStates {
+		wg.Add(1)
+		go func(as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) {
+			for _, tx := range as.FetchTxs(states, filterFn) {
+				etx := ms.deepCopyTx(tx)
+				txsLock.Lock()
+				txs = append(txs, etx)
+				txsLock.Unlock()
+			}
+			wg.Done()
+		}(as)
+	}
+	wg.Wait()
+
+	return txs, nil
+}
+
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) FindTxesWithMetaFieldByReceiptBlockNum(ctx context.Context, metaField string, blockNum int64, chainID *big.Int) ([]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], error) {
+	if ms.chainID.String() != chainID.String() {
+		return nil, fmt.Errorf("find_txes_with_meta_field_by_receipt_block_num: %w", ErrInvalidChainID)
+	}
+
+	filterFn := func(tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) bool {
+		if tx.Meta == nil {
+			return false
+		}
+		meta := map[string]interface{}{}
+		if err := json.Unmarshal(json.RawMessage(*tx.Meta), &meta); err != nil {
+			return false
+		}
+		if _, ok := meta[metaField]; !ok {
+			return false
+		}
+		if tx.TxAttempts == nil || len(tx.TxAttempts) == 0 {
+			return false
+		}
+		attempt := tx.TxAttempts[0]
```

Comment:
Shouldn't we traverse all of the attempts to find the one that has receipts?

Reply:
Yes, any attempt could have the receipt